glib/src/value.rs

```diff
@@ -930,6 +936,33 @@ impl SetValueOptional for str {
     }
 }
 
+impl<'a> FromValue<'a> for Value {
+    unsafe fn from_value(value: &'a Value) -> Self {
+        from_glib_none(
+            gobject_ffi::g_value_get_boxed(value.to_glib_none().0) as *mut gobject_ffi::GValue
+        )
+    }
+}
+
+impl<'a> FromValueOptional<'a> for Value {
+    unsafe fn from_value_optional(value: &'a Value) -> Option<Self> {
+        Some(FromValue::from_value(value))
+    }
+}
+
+impl SetValue for Value {
+    unsafe fn set_value(value: &mut Value, this: &Self) {
+        gobject_ffi::g_value_set_boxed(value.to_glib_none_mut().0, this.to_glib_full() as _)
+    }
+}
+
+impl SetValueOptional for Value {
+    unsafe fn set_value_optional(value: &mut Value, this: Option<&Self>) {
+        let ptr = this.map_or_else(ptr::null, |t| t.to_glib_full() as _);
```

Comment:
Same as above here. For simplicity I would probably just go with

```rust
if let Some(this) = this {
    gobject_ffi::g_value_set_boxed(value.to_glib_none_mut().0, this.to_glib_none().0 as ffi::gpointer);
} else {
    gobject_ffi::g_value_set_boxed(value.to_glib_none_mut().0, ptr::null_mut());
}
```

Reply:
This pattern from above is common in other parts of gtk-rs, except it's an explicit `match` for the pointer rather than `map_or_else`. Why is what you wrote "simpler", just because `map_or_else` is an uncommon combinator?

Reply:
Because in theory you have to keep the stash returned by `this.to_glib_none()` alive and only then pass the `.0` to the function. So the correct way with `map_or_else()` would be

```rust
let ptr = this.map(|t| t.to_glib_none()); // so that it stays alive
let ptr = ptr.map_or_else(ptr::null, |t|  t.0 as ffi::gconstpointer);
gobject_ffi::g_value_set_boxed(value.to_glib_none_mut().0, ptr);
```

Reply:
But your first suggestion also has `to_glib_none()` in a temporary that is dropped before the call to `g_value_get_boxed` 😅

Reply:
No, because it's in the same statement the return value of `to_glib_none()` is only dropped afterwards :)

Reply:
Huh, I did not know that. Thanks for the explanation!

Reply:
Otherwise you couldn't e.g. do `foo(&*some_ref_cell.borrow())` for example.